File: lttng-scope/src/main/java/org/lttng/scope/views/timecontrol/TimeControl.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.views.timecontrol;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.context.ViewGroupContext;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.geometry.HPos;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.geometry.VPos;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import javafx.scene.text.FontWeight;
import org.lttng.scope.views.context.ViewGroupContextManager;

import java.util.Arrays;
import java.util.stream.Stream;

import static org.lttng.scope.views.timecontrol.TimestampConversion.tsToString;

public class TimeControl extends BorderPane {

    /* UI Text */
    private static final String LABEL_VISIBLE_TIME_RANGE = "Visible Time Range";
    private static final String LABEL_SELECTION_TIME_RANGE = "Selection Time Range";
    private static final String LABEL_PROJECT_RANGE = "Trace Project Time Range";
    private static final String LABEL_START = "Start";
    private static final String LABEL_END = "End";
    private static final String LABEL_SPAN = "Span (s)";



    private static final long MINIMUM_VISIBLE_RANGE = 10000;
    private static final Font TITLE_FONT = Font.font(null, FontWeight.BOLD, -1);
    private static final Insets GRID_PADDING = new Insets(10);

    private final ObjectProperty<ViewGroupContext> viewContextProperty = new SimpleObjectProperty<>(ViewGroupContextManager.getCurrent());

    private final TextField[] textFields;

    public TimeControl() {
        ViewGroupContext viewCtx = viewContextProperty.get();
        // TODO Handle case where this view's ViewContext can change

        TimeRangeTextFields visibleRangeFields = new TimeRangeTextFields(viewCtx.getCurrentProjectFullRange(), MINIMUM_VISIBLE_RANGE);
        TimeRangeTextFields selectionRangeFields = new TimeRangeTextFields(viewCtx.getCurrentProjectFullRange(), null);

        textFields = new TextField[] {
                visibleRangeFields.getStartTextField(),
                visibleRangeFields.getEndTextField(),
                visibleRangeFields.getDurationTextField(),
                selectionRangeFields.getStartTextField(),
                selectionRangeFields.getEndTextField(),
                selectionRangeFields.getDurationTextField()
        };
        Arrays.stream(textFields).forEach(tf -> {
            tf.setAlignment(Pos.CENTER_LEFT);
        });

        TextField[] projRangeTextFields = Stream.generate(() -> new TextField()).limit(3)
                .peek(tf -> {
                    tf.setEditable(false);
                    tf.setPrefWidth(220);
                    tf.setAlignment(Pos.CENTER_LEFT);
                    tf.setBackground(new Background(new BackgroundFill(Color.LIGHTGRAY, new CornerRadii(2), null)));
                })
                .toArray(TextField[]::new);

        Label[] topHeaderLabels = Stream.of(LABEL_START, LABEL_END, LABEL_SPAN)
                .map(text -> new Label(text))
                .peek(label -> {
                    label.setFont(TITLE_FONT);
                    GridPane.setHalignment(label, HPos.CENTER);
                })
                .toArray(Label[]::new);

        Label[] leftHeaderLabels = Stream.of(LABEL_VISIBLE_TIME_RANGE, LABEL_SELECTION_TIME_RANGE, LABEL_PROJECT_RANGE)
                .map(text -> new Label(text))
                .peek(label -> {
                    label.setFont(TITLE_FONT);
                    GridPane.setHalignment(label, HPos.RIGHT);
                    GridPane.setValignment(label, VPos.CENTER);
                })
                .toArray(Label[]::new);

        /* Setup the elements in the grid */
        GridPane grid = new GridPane();
        grid.setPadding(GRID_PADDING);
        grid.setHgap(2);
        grid.setVgap(2);

        grid.add(topHeaderLabels[0], 1, 0);
        grid.add(topHeaderLabels[1], 2, 0);
        grid.add(topHeaderLabels[2], 3, 0);

        grid.add(leftHeaderLabels[0], 0, 1);
        grid.add(textFields[0], 1, 1);
        grid.add(textFields[1], 2, 1);
        grid.add(textFields[2], 3, 1);

        grid.add(leftHeaderLabels[1], 0, 2);
        grid.add(textFields[3], 1, 2);
        grid.add(textFields[4], 2, 2);
        grid.add(textFields[5], 3, 2);

        grid.add(leftHeaderLabels[2], 0, 3);
        grid.add(projRangeTextFields[0], 1, 3);
        grid.add(projRangeTextFields[1], 2, 3);
        grid.add(projRangeTextFields[2], 3, 3);


        viewCtx.currentTraceProjectProperty().addListener((obs, oldVal, newVal) -> {
            /* Update the displayed trace's time range. */
            TimeRange projRange;
            if (newVal == null) {
               projRange = ViewGroupContext.UNINITIALIZED_RANGE;
            } else {
                projRange  = newVal.getFullRange();
            }
            projRangeTextFields[0].setText(tsToString(projRange.getStartTime()));
            projRangeTextFields[1].setText(tsToString(projRange.getEndTime()));
            projRangeTextFields[2].setText(tsToString(projRange.getDuration()));

            /* Update the text fields' limits */
            visibleRangeFields.setLimits(projRange);
            selectionRangeFields.setLimits(projRange);

            /* Read initial dynamic range values from the view context. */
            TimeRange visibleRange = viewCtx.getCurrentVisibleTimeRange();
            TimeRange selectionRange = viewCtx.getCurrentSelectionTimeRange();
            textFields[0].setText(tsToString(visibleRange.getStartTime()));
            textFields[1].setText(tsToString(visibleRange.getEndTime()));
            textFields[2].setText(tsToString(visibleRange.getDuration()));
            textFields[3].setText(tsToString(selectionRange.getStartTime()));
            textFields[4].setText(tsToString(selectionRange.getEndTime()));
            textFields[5].setText(tsToString(selectionRange.getDuration()));

            attachRangeListeners(viewCtx);

            /* Bind underlying properties */
            visibleRangeFields.timeRangeProperty().bindBidirectional(viewCtx.currentVisibleTimeRangeProperty());
            selectionRangeFields.timeRangeProperty().bindBidirectional(viewCtx.currentSelectionTimeRangeProperty());
        });

        setCenter(grid);
    }

    /**
     * Add listeners to update the values when the context's time ranges change.
     * Note we dot no use binds here, we do not want every single keystroke in the
     * text fields to update the view context values!
     */
    private void attachRangeListeners(ViewGroupContext ctx) {
        ctx.currentVisibleTimeRangeProperty().addListener((obs, oldVal, newVal) -> {
            textFields[0].setText(tsToString(newVal.getStartTime()));
            textFields[1].setText(tsToString(newVal.getEndTime()));
            textFields[2].setText(tsToString(newVal.getDuration()));
        });
        ctx.currentSelectionTimeRangeProperty().addListener((obs, oldVal, newVal) -> {
            textFields[3].setText(tsToString(newVal.getStartTime()));
            textFields[4].setText(tsToString(newVal.getEndTime()));
            textFields[5].setText(tsToString(newVal.getDuration()));
        });
    }

}

File: lttng-scope/src/main/kotlin/org/lttng/scope/views/timecontrol/TimeControl.kt
/*
 * Copyright (C) 2017-2018 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.views.timecontrol

import com.efficios.jabberwocky.context.ViewGroupContext
import com.efficios.jabberwocky.project.TraceProject
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.geometry.HPos
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.geometry.VPos
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.scene.text.FontWeight
import org.lttng.scope.application.ScopeOptions
import org.lttng.scope.common.TimestampFormat
import org.lttng.scope.views.context.ViewGroupContextManager

class TimeControl(private val viewContext: ViewGroupContext) : BorderPane() {

    companion object {
        /* UI Text */
        private const val LABEL_VISIBLE_TIME_RANGE = "Visible Time Range"
        private const val LABEL_SELECTION_TIME_RANGE = "Selection Time Range"
        private const val LABEL_PROJECT_RANGE = "Trace Project Time Range"
        private const val LABEL_START = "Start"
        private const val LABEL_END = "End"
        private const val LABEL_SPAN = "Span (s)"

        private const val MINIMUM_VISIBLE_RANGE = 10000L
        private val TITLE_FONT = Font.font(null, FontWeight.BOLD, -1.0)
        private val GRID_PADDING = Insets(10.0)
        private const val TIMESTAMPS_FIELDS_WIDTH = 250.0
        private const val SPAN_FIELDS_WIDTH = 200.0
    }

    private val textFields: Array<TextField>
    private val visibleRangeFields: TimeRangeTextFields
    private val selectionRangeFields: TimeRangeTextFields
    internal val projRangeTextFields: Array<TextField>


    init {
        visibleRangeFields = TimeRangeTextFields(viewContext.getCurrentProjectFullRange(), MINIMUM_VISIBLE_RANGE)
        selectionRangeFields = TimeRangeTextFields(viewContext.getCurrentProjectFullRange(), null)

        textFields = arrayOf(
                visibleRangeFields.startTextField,
                visibleRangeFields.endTextField,
                visibleRangeFields.durationTextField,
                selectionRangeFields.startTextField,
                selectionRangeFields.endTextField,
                selectionRangeFields.durationTextField
        )
        textFields.forEach { it.alignment = Pos.CENTER_RIGHT }

        projRangeTextFields = generateSequence { TextField() }.take(3)
                .onEach {
                    it.isEditable = false
                    it.alignment = Pos.CENTER_RIGHT
                    it.background = Background(BackgroundFill(Color.LIGHTGRAY, CornerRadii(2.0), null))
                }.toList().toTypedArray()
        projRangeTextFields[0].prefWidth = TIMESTAMPS_FIELDS_WIDTH
        projRangeTextFields[1].prefWidth = TIMESTAMPS_FIELDS_WIDTH
        projRangeTextFields[2].prefWidth = SPAN_FIELDS_WIDTH

        val topHeaderLabels = listOf(LABEL_START, LABEL_END, LABEL_SPAN)
                .map { Label(it) }
                .onEach {
                    it.font = TITLE_FONT
                    GridPane.setHalignment(it, HPos.CENTER)
                }.toTypedArray()

        val leftHeaderLabels = listOf(LABEL_VISIBLE_TIME_RANGE, LABEL_SELECTION_TIME_RANGE, LABEL_PROJECT_RANGE)
                .map { Label(it) }
                .onEach {
                    it.font = TITLE_FONT
                    GridPane.setHalignment(it, HPos.RIGHT)
                    GridPane.setValignment(it, VPos.CENTER)
                }.toTypedArray()


        /* Setup the elements in the grid */
        val grid = GridPane().apply {
            padding = GRID_PADDING
            hgap = 2.0
            vgap = 2.0

            add(topHeaderLabels[0], 1, 0)
            add(topHeaderLabels[1], 2, 0)
            add(topHeaderLabels[2], 3, 0)

            add(leftHeaderLabels[0], 0, 1)
            add(textFields[0], 1, 1)
            add(textFields[1], 2, 1)
            add(textFields[2], 3, 1)

            add(leftHeaderLabels[1], 0, 2)
            add(textFields[3], 1, 2)
            add(textFields[4], 2, 2)
            add(textFields[5], 3, 2)

            add(leftHeaderLabels[2], 0, 3)
            add(projRangeTextFields[0], 1, 3)
            add(projRangeTextFields[1], 2, 3)
            add(projRangeTextFields[2], 3, 3)
        }

        center = grid
    }

    private val projectChangeListener = object : ViewGroupContext.ProjectChangeListener {
        override fun newProjectCb(newProject: TraceProject<*, *>?) {
            val tf = ScopeOptions.timestampFormat

            /* Update the displayed trace's time range. */
            val projRange = newProject?.fullRange ?: ViewGroupContext.UNINITIALIZED_RANGE

            /* Project ranges always use full [date]-[time] format, or s.ns if configured. */
            val projRangeFormat = when(tf) {
                TimestampFormat.HMS_N, TimestampFormat.YMD_HMS_N -> TimestampFormat.YMD_HMS_N
                TimestampFormat.SECONDS_POINT_NANOS -> TimestampFormat.SECONDS_POINT_NANOS
            }
            projRangeTextFields[0].text = projRangeFormat.tsToString(projRange.startTime)
            projRangeTextFields[1].text = projRangeFormat.tsToString(projRange.endTime)
            projRangeTextFields[2].text = TimestampFormat.SECONDS_POINT_NANOS.tsToString(projRange.duration)

            /* Update the text fields' limits */
            visibleRangeFields.limits = projRange
            selectionRangeFields.limits = projRange

            /* Read initial dynamic range values from the view context. */
            val visibleRange = viewContext.visibleTimeRange
            val selectionRange = viewContext.selectionTimeRange
            with(textFields) {
                get(0).text = tf.tsToString(visibleRange.startTime)
                get(1).text = tf.tsToString(visibleRange.endTime)
                get(2).text = TimestampFormat.SECONDS_POINT_NANOS.tsToString(visibleRange.duration)
                get(3).text = tf.tsToString(selectionRange.startTime)
                get(4).text = tf.tsToString(selectionRange.endTime)
                get(5).text = TimestampFormat.SECONDS_POINT_NANOS.tsToString(selectionRange.duration)
            }

            /**
             * Add listeners to update the values when the context's time ranges change.
             * Note we dot no use binds here, we do not want every single keystroke in the
             * text fields to update the view context values!
             */
            viewContext.visibleTimeRangeProperty().addListener { _, _, newVal ->
                if (viewContext.listenerFreeze) return@addListener
                textFields[0].text = tf.tsToString(newVal.startTime)
                textFields[1].text = tf.tsToString(newVal.endTime)
                textFields[2].text = TimestampFormat.SECONDS_POINT_NANOS.tsToString(newVal.duration)
            }
            viewContext.selectionTimeRangeProperty().addListener { _, _, newVal ->
                if (viewContext.listenerFreeze) return@addListener
                textFields[3].text = tf.tsToString(newVal.startTime)
                textFields[4].text = tf.tsToString(newVal.endTime)
                textFields[5].text = TimestampFormat.SECONDS_POINT_NANOS.tsToString(newVal.duration)
            }

            /* Bind underlying properties */
            visibleRangeFields.timeRangeProperty().bindBidirectional(viewContext.visibleTimeRangeProperty())
            selectionRangeFields.timeRangeProperty().bindBidirectional(viewContext.selectionTimeRangeProperty())
        }

    }

    private val timestampFormatChangeListener = ChangeListener<TimestampFormat> { _, _, newValue ->
        newValue ?: return@ChangeListener
        /*
         * The TimeRangeTextFields class already takes care of updating its contents
         * on format change. We only need to care care of the "project range" fields.
         */
        val formatToUse = when(newValue) {
            TimestampFormat.HMS_N, TimestampFormat.YMD_HMS_N -> TimestampFormat.YMD_HMS_N
            TimestampFormat.SECONDS_POINT_NANOS -> TimestampFormat.SECONDS_POINT_NANOS
        }
        val (start, end) = viewContext.getCurrentProjectFullRange()
        projRangeTextFields[0].text = formatToUse.tsToString(start)
        projRangeTextFields[1].text = formatToUse.tsToString(end)
        /*  Duration field remains at s.ns */
    }

    init {
        viewContext.registerProjectChangeListener(projectChangeListener)
        ScopeOptions.timestampFormatProperty().addListener(timestampFormatChangeListener)
    }

    @Suppress("ProtectedInFinal", "Unused")
    protected fun finalize() {
        viewContext.deregisterProjectChangeListener(projectChangeListener)
        ScopeOptions.timestampFormatProperty().removeListener(timestampFormatChangeListener)
    }

}
